examples/adt-tests.dx

```diff
@@ -209,3 +209,59 @@ def catLists (xs:List a) (ys:List a) : List a =
   n = 1 + 4
   AsList _ (for i:(Fin n). ordinal i)
 > (AsList 5 [0, 1, 2, 3, 4])
+
+
+
+def listToTable ((AsList n xs): List a) : (Fin n)=>a = xs
+
+:t listToTable
+> ((a:Type) ?-> (pat:(List a)) -> (Fin (%projectElt [0] pat)) => a)
```

Comment:
Hm, yeah that would look better, but it doesn't seem very easy to implement since that's not actually how it is represented in the IR. (And for other purposes we might want the IR pretty-printing to match the true IR structure.)

One compromise would be to do something like
```
((a:Type) ?-> (pat:(List a)) -> (Fin ((AsList n _) = pat; n)) => a)
```
or
```
((a:Type) ?-> (pat:(List a)) -> (Fin ((\(AsList n _). n) pat)) => a)
```
but those have their own downsides. Not sure what the best thing to do here is.

Reply:
I do agree that it's a bit annoying in that you need to inspect the latter parts of the type before you print it, but it would be the "desired end state". For now, at least doing the projection inline as you've suggested (both proposals looks ok to me) might be the way to go. Ideally those would also be valid user syntax, so that they can copy paste the output of `:t` into a type annotation. But it doesn't have to happen in this PR.

Reply:
I've implemented the second option, which should be valid user syntax even though it's a bit difficult to read.